amplitude.js

```diff
@@ -482,20 +517,30 @@ Amplitude.prototype.sendEvents = function() {
       upload_time: uploadTime,
       checksum: md5(API_VERSION + this.options.apiKey + events + uploadTime)
     };
-    var numEvents = this._unsentEvents.length;
+
     var scope = this;
-    new Request(url, data).send(function(response) {
+    new Request(url, data).send(function(status, response) {
       scope._sending = false;
       try {
-        if (response === 'success') {
+        if (status === 200 && response === 'success') {
           //log('sucessful upload');
-          scope._unsentEvents.splice(0, numEvents);
+          scope.removeEvents(maxEventId);
+
+          // Update the event cache after the removal of sent events.
           if (scope.options.saveEvents) {
             scope.saveEvents();
           }
+
+          // Send more events if any queued during previous send.
           if (scope._unsentEvents.length > 0) {
             scope.sendEvents();
           }
+        } else if (status === 413) {
+          //log('request too large');
+          // The server complained about the length of the request.
+          // Backoff and try again.
+          scope.options.uploadBatchSize = Math.floor(numEvents / 2);
```

Comment:
Is it most likely that the server is throwing an error because 1 event is massive or is it mostly when a client goes offline, tracks a ton of events and reconnects?

Maybe we should also put a limit on the size of any given event.